ceph-mon/lib/charms/ceph_mon/v0/ceph_cos_agent.py

```diff
@@ -0,0 +1,133 @@
+        )
+
+    def _on_refresh(self, event):
+        """Enable prometheus on relation change"""
+        if not ceph_utils.is_bootstrapped():
+            logger.debug("not bootstrapped, defer _on_refresh: %s", event)
+            event.defer()
+            return
+        logger.debug("refreshing cos_agent relation")
+        if self._charm.unit.is_leader():
+            self.mgr_config_set_rbd_stats_pools()
+            ceph_utils.mgr_enable_module("prometheus")
+        super()._on_refresh(event)
+
+    def _on_relation_departed(self, event):
+        """Disable prometheus on depart of relation"""
+        if self._charm.unit.is_leader() and ceph_utils.is_bootstrapped():
+            logger.debug(
+                "is_leader and is_bootstrapped, running rel departed: %s",
+                event,
+            )
+            ceph_utils.mgr_disable_module("prometheus")
+            logger.debug("module_disabled")
+
+    def _custom_scrape_configs(self):
+        fqdn = socket.getfqdn()
+        fqdn_parts = fqdn.split('.')
+        domain = '.'.join(fqdn_parts[1:]) if len(fqdn_parts) > 1 else fqdn
+        return [
+            {
+                "metrics_path": "/metrics",
+                "static_configs": [{"targets": ["localhost:9283"]}],
+                "honor_labels": True,
+                "metric_relabel_configs": [
+                    {
+                        # localhost:9283 is the generic default instance label
+                        # added by grafana-agent which is kinda useless.
+                        # Replace it with a somewhat more meaningful label
+                        "source_labels": ["instance"],
+                        "regex": "^localhost:9283$",
+                        "target_label": "instance",
+                        "action": "replace",
+                        "replacement": "ceph_cluster",
+                    },
+                    {   # if we have a non-empty hostname label, use it as the
+                        # instance label
+                        "source_labels": ["hostname"],
+                        "regex": "(.+)",
+                        "target_label": "instance",
+                        "action": "replace",
+                        "replacement": "${1}",
+                    },
+                    {   # tack on the domain to the instance label to make it
+                        # conform to grafana-agent's node-exporter expectations
+                        "source_labels": ["instance"],
+                        "regex": "(.*)",
+                        "target_label": "instance",
+                        "action": "replace",
+                        "replacement": "${1}." + domain,
+                    },
+                ]
+            },
+
+        ]
+
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(multiplier=1, max=10),
+        reraise=True,
+        stop=tenacity.stop_after_attempt(30))
+    def mgr_config_set_rbd_stats_pools(self):
```

Comment:
Hm, this function seems duplicated from the utils module. Would it make sense to move it to charms_ceph to reduce code duplication?

Reply:
added a TODO: on line number 109.